crates/rethnet_eth/src/remote/client.rs

```diff
@@ -376,12 +380,13 @@ impl RpcClient {
             })
             .collect::<Vec<_>>();
 
-        let mut results: Vec<Option<serde_json::Value>> =
-            Vec::with_capacity(method_invocations.len());
-        for cache_key in cache_keys.iter() {
-            let result = self.try_from_cache(cache_key.as_ref()).await?;
-            results.push(result)
-        }
+        let mut results: Vec<Option<serde_json::Value>> = stream::iter(&cache_keys)
+            .map(|cache_key| async { self.try_from_cache(cache_key.as_ref()).await })
+            .buffered(PARALLEL_DISK_READS)
```

Comment:
Should this be using something like https://doc.rust-lang.org/std/thread/fn.available_parallelism.html to be system-specific?

Reply:
This is about the number of futures we allow to be executed concurrently and as a result the max number of parallel disk reads we allow. I updated the code to use concurrent instead of parallel in the naming.

Reply:
I might be missing something, but that would still benefit from using a system-adapted parallelism or concurrency limit. If you're using more parallel readers than a system's multi-thread setup natively supports, you'll be unnecessarily saturating the runtime.

Reply:
I looked into this some more and I think it should be ok:

The [available_parallelism](https://doc.rust-lang.org/std/thread/fn.available_parallelism.html) function tells us the number of logical cores on the machine. This is the number of threads that can perform computation in parallel in a program and our Tokio runtime will spawn [this many workers threads](https://docs.rs/tokio/latest/tokio/runtime/struct.Builder.html#method.worker_threads) that are responsible for progressing futures. So if we block in async code, we're blocking the worker threads and there is indeed a problem.

But since my refactoring we're using `tokio::fs` instead of `std::fs` for file system i/o in the RPC client. These both perform blocking i/o for portability, but `tokio::fs` will spawn the blocking operations on a special thread pool for blocking operations and return a future that can be awaited without blocking the worker threads. The [default value](https://docs.rs/tokio/latest/tokio/runtime/struct.Builder.html#method.max_blocking_threads) for the size of the blocking thread pool is 512, so spawning max 16 concurrent blocking threads here keeps us well within this limit. The default size can be so large, because the various operating systems are very efficient at handling threads that are blocking on some syscall for i/o.

My concern with picking the `PARALLEL_DISK_READS` (since renamed to `CONCURRENT_DISK_READS`) was not so much about saturating the blocking thread pool, but  about performance degradation of disk throughput when issuing too many parallel reads. Based on the article [linked](https://pkolaczk.github.io/disk-parallelism/) in the comment 16 parallel disk reads is a very conservative parameter.